test/gc/transaction_level_gc_manager_test.cpp

```diff
@@ -368,123 +1225,15 @@ TEST_F(TransactionLevelGCManagerTests, ReInsertTest) {
-  auto tile_group = (table.get())->GetTileGroup(0);
-  auto tile_group_ptr = tile_group.get();
-  auto tile_group_header = tile_group_ptr->GetHeader();
-  tile_group_header->SetImmutability();
-
-  // Deleting a tuple from the 1st tilegroup
-  auto ret = DeleteTuple(table.get(), 2);
-  EXPECT_TRUE(ret == ResultType::SUCCESS);
-  epoch_manager.SetCurrentEpochId(2);
-  auto expired_eid = epoch_manager.GetExpiredEpochId();
-  EXPECT_EQ(1, expired_eid);
-  auto current_eid = epoch_manager.GetCurrentEpochId();
-  EXPECT_EQ(2, current_eid);
-  auto reclaimed_count = gc_manager.Reclaim(0, expired_eid);
-  auto unlinked_count = gc_manager.Unlink(0, expired_eid);
-  EXPECT_EQ(0, reclaimed_count);
-  EXPECT_EQ(1, unlinked_count);
-
-  epoch_manager.SetCurrentEpochId(3);
-  expired_eid = epoch_manager.GetExpiredEpochId();
-  EXPECT_EQ(2, expired_eid);
-  current_eid = epoch_manager.GetCurrentEpochId();
-  EXPECT_EQ(3, current_eid);
-  reclaimed_count = gc_manager.Reclaim(0, expired_eid);
-  unlinked_count = gc_manager.Unlink(0, expired_eid);
-  EXPECT_EQ(1, reclaimed_count);
-  EXPECT_EQ(0, unlinked_count);
-
-  // ReturnFreeSlot() should return null because deleted tuple was from
-  // immutable tilegroup.
-  auto location = gc_manager.ReturnFreeSlot((table.get())->GetOid());
-  EXPECT_EQ(location.IsNull(), true);
-
-  // Deleting a tuple from the 2nd tilegroup which is mutable.
-  ret = DeleteTuple(table.get(), 6);
-
-  EXPECT_TRUE(ret == ResultType::SUCCESS);
-  epoch_manager.SetCurrentEpochId(4);
-  expired_eid = epoch_manager.GetExpiredEpochId();
-  EXPECT_EQ(3, expired_eid);
-  current_eid = epoch_manager.GetCurrentEpochId();
-  EXPECT_EQ(4, current_eid);
-  reclaimed_count = gc_manager.Reclaim(0, expired_eid);
-  unlinked_count = gc_manager.Unlink(0, expired_eid);
-  EXPECT_EQ(0, reclaimed_count);
-  EXPECT_EQ(1, unlinked_count);
-
-  epoch_manager.SetCurrentEpochId(5);
-  expired_eid = epoch_manager.GetExpiredEpochId();
-  EXPECT_EQ(4, expired_eid);
-  current_eid = epoch_manager.GetCurrentEpochId();
-  EXPECT_EQ(5, current_eid);
-  reclaimed_count = gc_manager.Reclaim(0, expired_eid);
-  unlinked_count = gc_manager.Unlink(0, expired_eid);
-  EXPECT_EQ(1, reclaimed_count);
-  EXPECT_EQ(0, unlinked_count);
-
-  // ReturnFreeSlot() should not return null because deleted tuple was from
-  // mutable tilegroup.
-  location = gc_manager.ReturnFreeSlot((table.get())->GetOid());
-  EXPECT_EQ(location.IsNull(), false);
-
-  gc_manager.StopGC();
-  gc::GCManagerFactory::Configure(0);
-
-  table.release();
-  // DROP!
-  TestingExecutorUtil::DeleteDatabase("ImmutabilityDB");
-
-  auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
-  auto txn = txn_manager.BeginTransaction();
-  EXPECT_THROW(
-      catalog::Catalog::GetInstance()->GetDatabaseObject("ImmutabilityDB", txn),
-      CatalogException);
-  txn_manager.CommitTransaction(txn);
-}
+// TODO: add an immutability test back in, old one was not valid because it
```

Comment:
Identity theft like this is punishable and can lead to imprisonment. Ref:[https://www.experian.com/blogs/ask-experian/identity-theft-can-land-a-fraudster-in-jail/]. On a more serious note @poojanilangekar is right about her findings. If GC comes and wipes out the tuple that was representing the max value of the tilegroup, then the zone map is no longer accurate and hence this was added in to prevent GC from coming and messing around.

Reply:
Andy's explanation of the immutable flag to us was that the TileGroup could not be inserted, updated, or deleted from.